app/web_hub.go

```diff
@@ -687,6 +687,7 @@ func (i *hubConnectionIndex) RemoveInactiveConnections() {
 	now := model.GetMillis()
 	for conn := range i.byConnection {
 		if !conn.active && now-conn.lastUserActivityAt > i.staleThreshold.Milliseconds() {
+			conn.App.Srv().userService.ReturnSessionToPool(conn.GetSession())
```

Comment:
Do we actually need to return the session to the pool? Sessions for `WebConn` are never taken from the pool. I can imagine it may create more problems in the future as we add functionality.

Reply:
Sessions for webconn come from https://github.com/mattermost/mattermost-server/blob/93174fbe56f3d3dd16b97c84656c4a46305f31cd/api4/websocket.go#L43

which is actually taken from https://github.com/mattermost/mattermost-server/blob/93174fbe56f3d3dd16b97c84656c4a46305f31cd/web/handlers.go#L253

So they are originally taken from the pool, and then cloned.

It doesn't matter much I agree, but curious to know what sort of problems do you think might happen in future?

Reply:
@noxer - just checking - did you have any thoughts on this comment?

Reply:
>  just checking - did you have any thoughts on this comment?

Oh, sorry. Missed this one.

> So they are originally taken from the pool, and then cloned.
It doesn't matter much I agree, but curious to know what sort of problems do you think might happen in future?

As we've seen with this bug, other components access the same map. Even though the session had been returned to the pool. I don't seen anywhere that the map or other information is explicitly discarded.

Also -- as you mentioned -- this is a clone of the original session. So we feed back into the pool despite not using it directly.

I'm no more than 1/5 on this. When we're careful, it will probably be fine.

Reply:
>  I don't seen anywhere that the map or other information is explicitly discarded.

The webConn is being deleted from the connIndex. So it is going out of scope.

> So we feed back into the pool despite not using it directly.

Yes, this is only for the case of websocket requests though. For normal API requests, it is not cloned.

> I'm no more than 1/5 on this. When we're careful, it will probably be fine.

I agree. I don't think there is anything fundamentally wrong here, just a bug that we missed. But if we continue to see issues, I'll remove the line.